src/lexer.coffee

```diff
@@ -1264,7 +1264,7 @@ LINE_CONTINUER      = /// ^ \s* (?: , | \??\.(?![.\d]) | \??:: ) ///
 STRING_INVALID_ESCAPE = ///
   ( (?:^|[^\\]) (?:\\\\)* )        # Make sure the escape isn’t escaped.
   \\ (
-     ?: (0[0-7]|[1-7])             # octal escape
+     ?: (0\d|[1-7])                # octal escape
```

Comment:
> As far as I remember, this code is here to make sure that `"\0#{}7"` compiles to `` `\x007` ``. But id didn't seem to work in [try CoffeeScript](http://coffeescript.org/#try:%22%5C0%23%7B%7D7%22).
> 
> Either way, this should probably deal with `\08`, `\08` and `\09` as well, since they are invalid in strict mode and template literals.

Ok so it turns out this was here to guard against things like
```
"\0\
  7"
```
So extended these regexen per your suggestion to guard against `\00`/`\08`/`\09` as well

But your case `"\0#{}7"` is also good (and as you saw, currently generates invalid JS `` `\07` ``)

So my fix for that case was to add a similar guard to convert a trailing null escape (`\0`) to `\x00` in the `unquotedValueForTemplateLiteral` version of a `StringLiteral` (added tests cases)

It looks like we don't need a similar guard for the corresponding JSX case (eg `<div>\0{}8</div>`) since eg `<div>\08</div>` seems to be allowed and have its text content parsed as: a backslash (`\`), a zero (`0`), and an eight (`8`), not an (illegal) octal escape